File: app/src/main/java/com/lemonade/dima/screen/MessagesPresenter.java
package com.lemonade.dima.screen;

import com.lemonade.dima.model.Message;
import com.lemonade.dima.repository.LemonadeRepository;

import javax.inject.Inject;

import static com.lemonade.dima.screen.MessagesActivity.START;

public class MessagesPresenter {
    private final MessagesView messagesView;
    private LemonadeRepository repository;

    @Inject
    MessagesPresenter(MessagesView messagesView, LemonadeRepository repository) {
        this.messagesView = messagesView;
        this.repository = repository;
    }

    public void send(Message message) {
        if (!message.getText().equals(START))
            messagesView.addMessage(message);
        repository
                .message(message)
                .subscribe(messagesView::addMessages,
                        Throwable::printStackTrace);
    }

    public void savedMessages() {
        repository
                .saved()
                .subscribe(messages -> {
                    for (int i = 0; i < 2; i++)
                        messages.remove(messages.size() - 1);
                    if (messages.size() != 0)
                        messagesView.addMessagesToEnd(messages);
                });
    }
}

File: app/src/main/java/com/k4dima/chat/messages/MessagesPresenter.kt
package com.k4dima.chat.messages

import com.k4dima.chat.app.model.Message
import com.k4dima.chat.messages.MessagesActivity.Companion.START
import com.k4dima.chat.messages.repository.LemonadeRepository
import javax.inject.Inject

class MessagesPresenter
@Inject
internal constructor(private val messagesView: MessagesView,
                     private val repository: LemonadeRepository) {
    fun send(message: Message) {
        if (message.text != START)
            messagesView.addMessage(message)
        repository.message(message)
                .subscribe(messagesView::addMessages, Throwable::printStackTrace)
    }

    fun savedMessages() = repository.saved()
            .subscribe { messages ->
                messages.dropLast(2).run {
                    if (isNotEmpty())
                        messagesView.addMessagesToEnd(this)
                }
            }!!
}